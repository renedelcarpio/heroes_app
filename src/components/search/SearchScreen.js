import './SearchScreen.scss';

import { useLocation, useNavigate } from 'react-router-dom';

import HeroCard from '../hero/HeroCard';
import { getHeroesByName } from '../../helpers/getHeroesByName';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { useMemo } from 'react';

const SearchScreen = () => {
	const navigate = useNavigate();
	const location = useLocation();

	const { q = '' } = queryString.parse(location.search);

	const [value, handleInputChange] = useForm({
		searchText: q,
	});

	const { searchText } = value;

	const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

	const handleSearch = (e) => {
		e.preventDefault();
		navigate(`?q=${searchText}`);
	};

	return (
		<>
			<div className='search__container'>
				<form onSubmit={handleSearch}>
					<input
						type='text'
						placeholder='Search your favorite superhero'
						className='search__input'
						name='searchText'
						value={searchText}
						onChange={handleInputChange}
					/>
					<button className='search__button' type='submit'>
						Search...
					</button>
				</form>
				<div className='filtered--hero'>
					{!q ? (
						<h1 className='empty--hero'>Your heroes will appear here</h1>
					) : (
						heroesFiltered.length === 0 && (
							<h1 className='empty--hero'>There is no results for: {q}</h1>
						)
					)}
					{heroesFiltered.map((hero) => (
						<HeroCard key={hero.id} {...hero} />
					))}
				</div>
			</div>
		</>
	);
};

export default SearchScreen;
