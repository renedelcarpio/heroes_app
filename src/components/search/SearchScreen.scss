@import '../../index.scss';

.search {
  &__container {
    background-color: $color-dark;
    display: flex;
    flex-direction: column;
    align-items: center;
    padding-top: 2rem;
    width: 100vw;
    min-height: 100vh;
  }

  &__input {
    background-color: rgba($color-mid, 0.3);
    border: none;
    border-radius: 2rem;
    color: $color-dark;
    margin-right: 1rem;
    padding: 1rem 2rem;
    outline: none;
  }

  &__button {
    background-color: transparent;
    border: 0.2rem solid $color-primary-dc;
    border-radius: 0.5rem;
    color: $color-primary-dc;
    font-size: 1.3rem;
    font-weight: 400;
    padding: 0.3rem 0.6rem;
    transition: all 0.3s;

    &:hover {
      background-color: $color-primary-dc;
      color: $color-light;
    }
  }
}

.filtered--hero {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.empty--hero {
  color: $color-light;
  margin-top: 5rem;
}
