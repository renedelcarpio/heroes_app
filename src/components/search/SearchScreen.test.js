import { MemoryRouter } from 'react-router-dom';
import SearchScreen from './SearchScreen';
import { mount } from 'enzyme';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockNavigate,
}));

describe('Tests in SearchScreen component', () => {
	test('should match with snapshot', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<SearchScreen />
			</MemoryRouter>
		);

		expect(wrapper).toMatchSnapshot();
		expect(typeof wrapper.find('.empty--hero')).toEqual('object');
		//TODO: Review why this test does not work
		// expect(wrapper.find('.empty--hero')).toEqual(
		// 	'Your heroes will appear here'
		// );
	});

	test('should show Batman and the input with the query string value', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman']}>
				<SearchScreen />
			</MemoryRouter>
		);

		expect(wrapper.find('input').prop('value')).toBe('batman');
	});

	test('should show an error message if query does not match with some hero', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=Batman123']}>
				<SearchScreen />
			</MemoryRouter>
		);

		expect(typeof wrapper.find('.empty--hero')).toEqual('object');
		//TODO: Review why this test does not work
		// expect(wrapper.find('.empty--hero')).toEqual(
		// 	'There is no results for: Batman123'
		// );
	});

	test('should call navigate to new screen', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<SearchScreen />
			</MemoryRouter>
		);

		wrapper.find('input').simulate('change', {
			target: {
				name: 'searchText',
				value: 'ironman',
			},
		});

		wrapper.find('form').prop('onSubmit')({
			preventDefault: () => {},
		});

		expect(mockNavigate).toHaveBeenCalledWith('?q=ironman');
	});
});
