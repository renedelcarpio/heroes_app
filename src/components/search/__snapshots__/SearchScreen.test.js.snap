// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in SearchScreen component should match with snapshot 1`] = `
<div
  className="search__container"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="search__input"
      name="searchText"
      onChange={[Function]}
      placeholder="Search your favorite superhero"
      type="text"
      value=""
    />
    <button
      className="search__button"
      type="submit"
    >
      Search...
    </button>
  </form>
  <div
    className="filtered--hero"
  >
    <h1
      className="empty--hero"
    >
      Your heroes will appear here
    </h1>
  </div>
</div>
`;
